src/sema/builtin.rs

```diff
@@ -1103,10 +1103,10 @@ pub(super) fn resolve_namespace_call(
                             broken = true;
                         }
 
-                        if ty.is_mapping() {
+                        if ty.is_mapping() || ty.is_recursive(ns) {
                             diagnostics.push(Diagnostic::error(
                                 *loc,
-                                "mapping cannot be abi decoded or encoded".to_string(),
+                                format!("'{}' cannot be abi decoded or encoded", ty.to_string(ns)),
```

Comment:
It would be nice if the error message explained that the type cannot be encoded because it's recursive.